src/basic/unit-name.c

```diff
@@ -507,6 +509,42 @@ int unit_name_template(const char *f, char **ret) {
         return 0;
 }
 
+int unit_name_hash_long(const char *name, char **ret) {
+        _cleanup_free_ char *s = NULL;
+        char hash[17] = {}, n[UNIT_NAME_MAX] = {}, *suffix;
+        uint64_t h;
+        uint8_t k[16] = {};
+        int i;
+
+        if (strlen(name) < UNIT_NAME_MAX)
+                return 0;
+
+        suffix = strrchr(name, '.');
+        if (!suffix)
+                return -EINVAL;
+
+        if (unit_type_from_string(suffix+1) < 0)
+                return -EINVAL;
+
+        /* We intentionally hash with k that is all zeros because each time we want the same hash value for the same string. */
+        h = siphash24_string(name, k);
+
+        for (i = 0; i < 8; i++) {
+                uint64_t m = (uint64_t) 0xff << 8 * i;
+                sprintf(&hash[i*2], "%02hhx", (char) ((h & m) >> 8*i));
+        }
+
+        stpcpy(stpcpy(stpcpy(mempcpy(n, name, UNIT_NAME_MAX - 1 - strlen(suffix+1) - sizeof(hash) - 1 - 3), "..."), hash), suffix);
```

Comment:
strjoin()?

Reply:
i think adding a marker for hashed names makes sense, but i'd probably use something shorter, maybe just "_"?